// appium-xcuitest-driver/build/lib/commands/log.js
'use strict';

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumIosDriver = require('appium-ios-driver');

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var extensions = {};

_Object$assign(extensions, _appiumIosDriver.iosCommands.logging);

extensions.startLogCapture = function callee$0$0() {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        this.logs = this.logs || {};

        if (_lodash2['default'].isEmpty(this.logs)) {
          context$1$0.next = 4;
          break;
        }

        _logger2['default'].warn('Trying to start iOS log capture but it has already started!');
        return context$1$0.abrupt('return');

      case 4:
        this.logs.crashlog = new _appiumIosDriver.IOSCrashLog();
        this.logs.syslog = new _appiumIosDriver.IOSLog({
          sim: this.opts.device,
          udid: this.isRealDevice() ? this.opts.udid : undefined,
          showLogs: this.opts.showIOSLog,
          realDeviceLogger: this.opts.realDeviceLogger
        });
        context$1$0.prev = 6;
        context$1$0.next = 9;
        return _regeneratorRuntime.awrap(this.logs.syslog.startCapture());

      case 9:
        context$1$0.next = 16;
        break;

      case 11:
        context$1$0.prev = 11;
        context$1$0.t0 = context$1$0['catch'](6);

        _logger2['default'].warn('Could not capture logs from device: ' + context$1$0.t0.message);
        _logger2['default'].debug('Continuing without capturing logs.');
        return context$1$0.abrupt('return');

      case 16:
        context$1$0.next = 18;
        return _regeneratorRuntime.awrap(this.logs.crashlog.startCapture());

      case 18:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[6, 11]]);
};

exports['default'] = extensions;
module.exports = exports['default'];
