// appium-xcuitest-driver/build/lib/commands/screenshots.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _asyncbox = require('asyncbox');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _nodeSimctl = require('node-simctl');

var _teen_process = require('teen_process');

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var _appiumSupport = require('appium-support');

var commands = {};

function getScreenshotWithIdevicelib(udid) {
  var pathToScreenshotTiff, pathToResultPng;
  return _regeneratorRuntime.async(function getScreenshotWithIdevicelib$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(_appiumSupport.tempDir.path({ prefix: 'screenshot-' + udid, suffix: '.tiff' }));

      case 2:
        pathToScreenshotTiff = context$1$0.sent;
        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.rimraf(pathToScreenshotTiff));

      case 5:
        context$1$0.next = 7;
        return _regeneratorRuntime.awrap(_appiumSupport.tempDir.path({ prefix: 'screenshot-' + udid, suffix: '.png' }));

      case 7:
        pathToResultPng = context$1$0.sent;
        context$1$0.next = 10;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.rimraf(pathToResultPng));

      case 10:
        context$1$0.prev = 10;
        context$1$0.prev = 11;
        context$1$0.next = 14;
        return _regeneratorRuntime.awrap((0, _teen_process.exec)('idevicescreenshot', ['-u', udid, pathToScreenshotTiff]));

      case 14:
        context$1$0.next = 20;
        break;

      case 16:
        context$1$0.prev = 16;
        context$1$0.t0 = context$1$0['catch'](11);

        _logger2['default'].warn('Cannot take a screenshot from the device ' + udid + ' using idevicescreenshot. Original error: ' + context$1$0.t0.message);
        return context$1$0.abrupt('return');

      case 20:
        context$1$0.prev = 20;
        context$1$0.next = 23;
        return _regeneratorRuntime.awrap((0, _teen_process.exec)('sips', ['-s', 'format', 'png', pathToScreenshotTiff, '--out', pathToResultPng]));

      case 23:
        context$1$0.next = 29;
        break;

      case 25:
        context$1$0.prev = 25;
        context$1$0.t1 = context$1$0['catch'](20);

        _logger2['default'].warn('Cannot convert a screenshot from TIFF to PNG using sips tool. Original error: ' + context$1$0.t1.message);
        return context$1$0.abrupt('return');

      case 29:
        context$1$0.next = 31;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(pathToResultPng));

      case 31:
        if (context$1$0.sent) {
          context$1$0.next = 34;
          break;
        }

        _logger2['default'].warn('Cannot convert a screenshot from TIFF to PNG. The conversion result does not exist at ' + pathToResultPng);
        return context$1$0.abrupt('return');

      case 34:
        context$1$0.next = 36;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.readFile(pathToResultPng));

      case 36:
        return context$1$0.abrupt('return', context$1$0.sent.toString('base64'));

      case 37:
        context$1$0.prev = 37;
        context$1$0.next = 40;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.rimraf(pathToScreenshotTiff));

      case 40:
        context$1$0.next = 42;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.rimraf(pathToResultPng));

      case 42:
        return context$1$0.finish(37);

      case 43:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[10,, 37, 43], [11, 16], [20, 25]]);
}

commands.getScreenshot = function callee$0$0() {
  var getScreenshotFromWDA, data, result;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    var _this = this;

    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        getScreenshotFromWDA = function getScreenshotFromWDA() {
          var data;
          return _regeneratorRuntime.async(function getScreenshotFromWDA$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                context$2$0.next = 2;
                return _regeneratorRuntime.awrap(this.proxyCommand('/screenshot', 'GET'));

              case 2:
                data = context$2$0.sent;

                if (_lodash2['default'].isString(data)) {
                  context$2$0.next = 5;
                  break;
                }

                throw new Error('Unable to take screenshot. WDA returned \'' + JSON.stringify(data) + '\'');

              case 5:
                return context$2$0.abrupt('return', data);

              case 6:
              case 'end':
                return context$2$0.stop();
            }
          }, null, _this);
        };

        context$1$0.prev = 1;
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(getScreenshotFromWDA());

      case 4:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 7:
        context$1$0.prev = 7;
        context$1$0.t0 = context$1$0['catch'](1);

        if (!(!this.isRealDevice() && this.xcodeVersion.versionFloat >= 8.1)) {
          context$1$0.next = 14;
          break;
        }

        _logger2['default'].info('Falling back to \'simctl io screenshot\' API');
        context$1$0.next = 13;
        return _regeneratorRuntime.awrap((0, _nodeSimctl.getScreenshot)(this.opts.udid));

      case 13:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 14:
        context$1$0.t1 = this.isRealDevice();

        if (!context$1$0.t1) {
          context$1$0.next = 19;
          break;
        }

        context$1$0.next = 18;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.which('idevicescreenshot'));

      case 18:
        context$1$0.t1 = context$1$0.sent;

      case 19:
        if (!context$1$0.t1) {
          context$1$0.next = 26;
          break;
        }

        _logger2['default'].info('Falling back to \'idevicescreenshot\' API');
        context$1$0.next = 23;
        return _regeneratorRuntime.awrap(getScreenshotWithIdevicelib(this.opts.udid));

      case 23:
        data = context$1$0.sent;

        if (!data) {
          context$1$0.next = 26;
          break;
        }

        return context$1$0.abrupt('return', data);

      case 26:
        result = undefined;
        context$1$0.next = 29;
        return _regeneratorRuntime.awrap((0, _asyncbox.retryInterval)(5, 1000, function callee$2$0() {
          return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
            while (1) switch (context$3$0.prev = context$3$0.next) {
              case 0:
                context$3$0.next = 2;
                return _regeneratorRuntime.awrap(getScreenshotFromWDA());

              case 2:
                result = context$3$0.sent;

              case 3:
              case 'end':
                return context$3$0.stop();
            }
          }, null, _this);
        }));

      case 29:
        return context$1$0.abrupt('return', result);

      case 30:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[1, 7]]);
};

exports['default'] = commands;
module.exports = exports['default'];

// The sips tool is only present on Mac OS

// Retry for real devices only. Fail fast on Simulator if simctl does not work as expected
