// appium-ios-driver/build/test/e2e/setup-base.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _helpersEnv = require('./helpers/env');

var _helpersEnv2 = _interopRequireDefault(_helpersEnv);

var _helpersSession = require('./helpers/session');

var _helpersTitle = require('./helpers/title');

var _appiumBaseDriver = require('appium-base-driver');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _libLogger = require('../../lib/logger');

var _libLogger2 = _interopRequireDefault(_libLogger);

require('./helpers/setup_testlibs');

var server = undefined;

function startServer(session) {
  var router;
  return _regeneratorRuntime.async(function startServer$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        router = (0, _appiumBaseDriver.routeConfiguringFunction)(session.rawDriver);
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap((0, _appiumBaseDriver.server)(router, _helpersEnv2['default'].APPIUM_PORT, 'localhost'));

      case 3:
        server = context$1$0.sent;

        _libLogger2['default'].info('IosDriver server listening on http://localhost:' + _helpersEnv2['default'].APPIUM_PORT);

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function setup(context, desired) {
  var opts = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

  var _this = this;

  var envOverrides = arguments.length <= 3 || arguments[3] === undefined ? false : arguments[3];
  var needsNewServer = arguments.length <= 4 || arguments[4] === undefined ? false : arguments[4];

  context.timeout(_helpersEnv2['default'].MOCHA_INIT_TIMEOUT);
  context.retries(2);
  var newEnv = _lodash2['default'].clone(_helpersEnv2['default']);
  if (envOverrides) {
    _lodash2['default'].extend(newEnv, envOverrides);
  }

  if (!opts.port) {
    opts.port = _helpersEnv2['default'].APPIUM_PORT;
  }

  opts.callbackAddress = 'localhost';

  var session = new _helpersSession.Session(desired, opts);
  var allPassed = true;

  before(function callee$1$0() {
    return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          if (server) {
            context$2$0.next = 5;
            break;
          }

          context$2$0.next = 3;
          return _regeneratorRuntime.awrap(startServer(session));

        case 3:
          context$2$0.next = 10;
          break;

        case 5:
          if (!needsNewServer) {
            context$2$0.next = 10;
            break;
          }

          context$2$0.next = 8;
          return _regeneratorRuntime.awrap(server.close());

        case 8:
          context$2$0.next = 10;
          return _regeneratorRuntime.awrap(startServer(session));

        case 10:
          context$2$0.next = 12;
          return _regeneratorRuntime.awrap(session.setUp((0, _helpersTitle.getTitle)(context)));

        case 12:
        case 'end':
          return context$2$0.stop();
      }
    }, null, _this);
  });

  after(function callee$1$0() {
    return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          context$2$0.next = 2;
          return _regeneratorRuntime.awrap(session.tearDown(allPassed));

        case 2:
          if (!needsNewServer) {
            context$2$0.next = 5;
            break;
          }

          context$2$0.next = 5;
          return _regeneratorRuntime.awrap(server.close());

        case 5:
        case 'end':
          return context$2$0.stop();
      }
    }, null, _this);
  });

  afterEach(function () {
    allPassed = allPassed && this.currentTest.state === 'passed';
  });

  return session;
}

exports['default'] = setup;
module.exports = exports['default'];

// start the server before start the session, so startup can use it if necessary
